ml-agents/mlagents/trainers/saver/tf_saver.py

```diff
@@ -0,0 +1,145 @@
+import os
+import shutil
+from typing import Optional
+from mlagents_envs.exception import UnityPolicyException
+from mlagents_envs.logging_util import get_logger
+from mlagents.tf_utils import tf
+from mlagents.trainers.saver.saver import BaseSaver
+from mlagents.trainers.tf.model_serialization import export_policy_model
+from mlagents.trainers.settings import TrainerSettings, SerializationSettings
+from mlagents.trainers.policy.tf_policy import TFPolicy
+from mlagents.trainers import __version__
+
+
+logger = get_logger(__name__)
+
+
+class TFSaver(BaseSaver):
+    """
+    Saver class for TensorFlow
+    """
+
+    def __init__(
+        self, trainer_settings: TrainerSettings, model_path: str, load: bool = False
+    ):
+        super().__init__()
+        self.model_path = model_path
+        self.initialize_path = trainer_settings.init_path
+        self._keep_checkpoints = trainer_settings.keep_checkpoints
+        self.load = load
+
+        # Currently only support saving one policy. This is the one to be saved.
+        self.policy: Optional[TFPolicy] = None
+
+    def register(self, module, init_or_load=True):
+        if isinstance(module, TFPolicy):
+            if self.policy is None:
+                self.policy = module
```

Comment:
Should it be an error if `self.policy` is not None? What about if `self.policy != module`?

Reply:
Also, should initialize_or_load() be called if self.policy wasn't None? Right now it looks like we could call initialize_or_load on a policy twice.

Reply:
To the first response, it's not an error. This line correspond to the warning in `rl_trainer.py`: `Trainer has multiple policies, but default behavior only saves the first.`. There might be cases we have multiple policies and others are just ignored for checkpointing. I can put some message here like `Registering multiple policies while the default behavior only saves the first. This model will not be saved.`

Reply:
It shouldn't be called twice. During add_policy, the policy is registered to the saver. At the same time, call initialize_or_load on that policy. Can you point me out where could it be called initialize_or_load twice?

Reply:
> Can you point me out where could it be called initialize_or_load twice?

This looks OK now. In the original version that I commented on, the `initialize_or_load` call was outside of the `if self.policy is None` block